3party/opening_hours/osm_time_range.hpp

```diff
@@ -27,108 +27,491 @@
+}
+
+std::ostream & operator<<(std::ostream & ost, Time::EEvent const event);
+std::ostream & operator<<(std::ostream & ost, Time const & time);
+
+class Timespan
+{
+ public:
+  Timespan() = default;
+  Timespan(Timespan const &) = default;
+  Timespan(Time const & start, bool plus = false);
+  Timespan(Time const & start, Time const & end, bool plus = false);
+  Timespan(Time const & start, Time const & end, Time const & period);
+
+  bool IsOpen() const;
+  bool HasPlus() const;
+  bool HasPeriod() const;
+
+  Time const & GetStart() const;
+  Time const & GetEnd() const;
+  Time const & GetPeriod() const;
+
+  void SetStart(Time const & start);
+  void SetEnd(Time const & end);
+  void SetPeriod(Time const & period);
+  void SetPlus(bool const plus);
+
+  bool IsValid() const;
+
+ private:
+  Time m_start;
+  Time m_end;
+  Time m_period;
+  bool m_plus{false};
+};
+
+using TTimespans = std::vector<Timespan>;
 
-  class State
+std::ostream & operator<<(std::ostream & ost, Timespan const & span);
+std::ostream & operator<<(std::ostream & ost, osmoh::TTimespans const & timespans);
+
+class NthEntry
+{
+ public:
+  enum class ENth
   {
-  public:
-    enum EState {
-      eUnknown = 0,
-      eClosed = 1,
-      eOpen = 2
-    };
+    None,
+    First,
+    Second,
+    Third,
+    Fourth,
+    Fifth
+  };
+
+ public:
+  bool IsEmpty() const;
+  bool HasStart() const;
+  bool HasEnd() const;
+
+  ENth GetStart() const;
+  ENth GetEnd() const;
+
+  void SetStart(ENth const s);
+  void SetEnd(ENth const e);
+
+ private:
+  ENth m_start{};
+  ENth m_end{};
+};
+
+std::ostream & operator<<(std::ostream & ost, NthEntry const entry);
+
+enum class EWeekday
+{
+  None,
+  Su,
+  Mo,
+  Tu,
+  We,
+  Th,
+  Fr,
+  Sa
+};
+
+inline constexpr EWeekday operator ""_day(unsigned long long day)
+{
+  using TDay = decltype(day);
+  return ((day <= static_cast<TDay>(EWeekday::None) ||
+           day > static_cast<TDay>(EWeekday::Sa))
+          ? EWeekday::None
+          : static_cast<EWeekday>(day));
+}
+
+std::ostream & operator<<(std::ostream & ost, EWeekday const wday);
+
+class WeekdayRange
+{
+  using TNths = std::vector<NthEntry>;
+
+ public:
+
+ public:
+  bool HasWday(EWeekday const & wday) const;
+
+  bool HasSu() const;
+  bool HasMo() const;
+  bool HasTu() const;
+  bool HasWe() const;
+  bool HasTh() const;
+  bool HasFr() const;
+  bool HasSa() const;
+
+  bool HasStart() const;
+  bool HasEnd() const;
+  bool IsEmpty() const;
+
+  EWeekday GetStart() const;
+  EWeekday GetEnd() const;
+  size_t GetDaysCount() const;
+
+  void SetStart(EWeekday const & wday);
+  void SetEnd(EWeekday const & wday);
+
+  int32_t GetOffset() const;
+  void SetOffset(int32_t const offset);
+
+  bool HasNth() const;
+  TNths const & GetNths() const;
+
+  void AddNth(NthEntry const & entry);
+
+ private:
+  EWeekday m_start{};
+  EWeekday m_end{};
+  int32_t m_offset{};
+  TNths m_nths;
+};
+
+using TWeekdayRanges = std::vector<WeekdayRange>;
```

Comment:
This shortcut is not too short and used only twice in this CL. I think we don't need it.